src/app/project/project.service.ts

```diff
@@ -1,142 +1,223 @@
-      project_id: 55,
-      progress: 0.62,
-      Activity :[
-        { 
-          id: 103,
-          name: 'Test Activity three',
-          isLocked: false,
-          progress: 0.47,
-          hasFeedback: false,
-        },
-        {
-          id: 104,
-          name: 'Activity four',
-          isLocked: false,
-          progress:0.98,
-          hasFeedback: false,
-        },
-        {
-          id: 105,
-          name: 'Test Activity five',
-          isLocked: false,
-          progress: 0.47,
-          hasFeedback: false,
-        },
-      ],
-      isLocked: false
-    },
-    {
-      id: 3,
-      name: 'communication',
-      description: 'there is some dummy text',
-      progress: 1,
-      project_id: 55,
-      Activity :[
-        {
-          id: 106,
-          name: 'Test Activity six',
-          isLocked: false,
-          progress: 1,
-          hasFeedback: false,
-        },
-        {
-          id: 107,
-          name: 'Activity seven',
-          isLocked: false,
-          progress:0.48,
-          hasFeedback: true,
+  public milestones: Array<Milestone> = [];
+  public milestone_ids: Array<number> = [];
+  public activities: Array<Activity> = [];
+  
+  constructor( 
+    private request: RequestService,
+    private utils: UtilsService) { }
+  
+  public getMilestones() {
+    return this.request.get(api.milestone)
+    .pipe(map(response => {
+      if (response.success && response.data) {
+        return this._normaliseMilestones(response.data);
+      }
+    }));
+  }
+  
+  private _normaliseMilestones(data){
+    
+    if (!Array.isArray(data)) {
+      this.request.apiResponseFormatError('Milestones array format error');
+      return [];
+    }
+    this.milestones = [];
+    this.milestone_ids = [];
+    data.forEach(eachMilestone => {
+      if (!this.utils.has(eachMilestone, 'id') || 
+          !this.utils.has(eachMilestone, 'name') || 
+          !this.utils.has(eachMilestone, 'is_locked')) {
+        return this.request.apiResponseFormatError('Milestone format error');
+      }
+      let milestone: Milestone = {
+        id: 0,
+        name: '',
+        description: '',
+        isLocked: false,
+        progress: 0,
+        Activity:[]
+      };
+      milestone.id = eachMilestone.id;
+      milestone.name = eachMilestone.name;
+      milestone.isLocked = eachMilestone.is_locked;
+      if (this.utils.has(eachMilestone, 'description')) {
+        milestone.description = eachMilestone.description;
+      };
+      
+      this.milestones.push(milestone);
+      
+  });
+  this.milestones.forEach(milestone => {
+    if (!milestone.isLocked) {
+      this.milestone_ids.push(milestone.id);
+    }
+  });
+
+  this._getActivities().subscribe(activities => { 
+    
+    this.activities = activities; 
+    this._addActivitiesToEachMilestone(this.milestones, this.activities);
+    this.milestone_ids.forEach(id => {
+      this._getProgress(id, this.milestones).subscribe();
```

Comment:
We don't need to get the progress within the loop of milestones. Otherwise we will call this API multiple times to get those progresses. Instead, we can just call it once to get all progresses. Also, the scope shouldn't be "Task" since we only need the progress of activities.